src/Integration/MefServices/ProjectsRuleSetProvider.cs

```diff
@@ -0,0 +1,355 @@
+            this.activeSolutionTracker.AfterProjectOpened += OnAfterProjectOpened;
+
+            BuildCacheAsync();
+        }
+
+        private async void OnAfterProjectOpened(object sender, ProjectOpenedEventArgs e)
+        {
+            if (this.projectSystemHelper.IsSolutionFullyOpened() &&
+                e.Project != null &&
+                !string.IsNullOrEmpty(e.Project.FullName) &&
+                !projectPathToCachedData.ContainsKey(e.Project.FullName))
+            {
+                // Improve me: there is no need to rebuild the full cache here
+                ClearCache();
+                await BuildCacheAsync();
+            }
+        }
+
+        private async void OnActiveSolutionChanged(object sender, ActiveSolutionChangedEventArgs e)
+        {
+            ClearCache();
+
+            if (e.IsSolutionOpen)
+            {
+                await BuildCacheAsync();
+            }
+        }
+
+        public bool HasRuleSetWithSonarAnalyzerRules(string projectFilePath) =>
+            projectPathToCachedData.ContainsKey(projectFilePath) && projectPathToCachedData[projectFilePath].HasAnySonarRule;
+
+        private void ClearCache()
+        {
+            filesToTrack.Values.ToList().ForEach(x => x.UpdatedOnDisk -= OnUpdatedOnDisk);
+            filesToTrack.Clear();
+
+            projectPathToCachedData.Clear();
+        }
+
+        private async Task BuildCacheAsync()
+        {
+            while (!projectSystemHelper.IsSolutionFullyOpened())
+            {
+                await Task.Delay(2000); // REVIEW: Is 2s a good enough time?
+            }
+
+            projectSystemHelper.GetSolutionProjects()
+                .ToList()
+                .ForEach(p => projectPathToCachedData.Add(p.FullName, ProcessProjectAndCollectFilesToTrack(p)));
+
+            // Subscribe to file changes for the projects and the rulesets
+            filesToTrack.Values.ToList().ForEach(x => x.UpdatedOnDisk += OnUpdatedOnDisk);
+
+            if (this.configurationProvider.GetConfiguration().Mode == SonarLintMode.Connected)
+            {
+                var projectsWithSonarRuleSet = projectPathToCachedData.Where(x => x.Value.HasAnySonarRule)
+                    .Select(x => Path.GetFileNameWithoutExtension(x.Key))
+                    .ToList();
+                if (projectsWithSonarRuleSet.Count > 0)
+                {
+                    this.logger.WriteLine(Strings.WarningProjectsWithSonarRulesWhileNewConnected,
+                        string.Join(", ", projectsWithSonarRuleSet));
+                }
+            }
+        }
+
+        private async void OnUpdatedOnDisk(object sender, EventArgs e)
+        {
+            var fileChangeTracker = sender as FileChangeTracker;
+            if (fileChangeTracker == null)
+            {
+                return;
+            }
+
+            if (!fileChangeTracker.FilePath.EndsWith(".ruleset"))
```

Comment:
I'm not sure if a ruleset _has_ to have a _.ruleset_ extension; I suspect VS and MSBuild would handle it correctly if it didn't, but I've never seen a ruleset file with a different extension. Not worth worrying about.

Reply:
I actually wanted to start with the other way (i.e. detecting we are called-back for a project but I don't have the info).